src/Features/CSharp/Portable/CodeRefactorings/SyncNamespace/CSharpSyncNamespaceCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,217 @@
+        {
+            var part = namespaceParts[index].EscapeIdentifier();
+            var namePiece = SyntaxFactory.IdentifierName(part);
+
+            if (index == 0)
+            {
+                return aliasQualifier == null ? (NameSyntax)namePiece : SyntaxFactory.AliasQualifiedName(aliasQualifier, namePiece);
+            }
+            else
+            {
+                return SyntaxFactory.QualifiedName(CreateNameSyntax(namespaceParts, aliasQualifier, index - 1), namePiece);
+            }                                                                              
+        }
+
+        protected override async Task<(bool, NamespaceDeclarationSyntax)> ShouldPositionTriggerRefactoringAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            
+            var namespaceDeclarationCount = root.DescendantNodes().OfType<NamespaceDeclarationSyntax>().Count();
+
+            if (namespaceDeclarationCount == 1)
+            {
+                var token = root.FindToken(position);
+                if (token.Kind() == SyntaxKind.EndOfFileToken)
+                {
+                    token = token.GetPreviousToken();
+                }
+                // Should trigger if cursor is on the name of only namespace declaration in this document.
+                var namespaceDeclaration = token.GetAncestor<NamespaceDeclarationSyntax>();
+
+                var shouldTrigger = namespaceDeclaration != null 
+                    && namespaceDeclaration.Name.Span.IntersectsWith(position) 
+                    && namespaceDeclaration.GetDiagnostics().All(diag => diag.Severity != DiagnosticSeverity.Error)
+                    && !ContainsPartialDeclaration(document, namespaceDeclaration, cancellationToken);
+                return (shouldTrigger, shouldTrigger ? namespaceDeclaration : null);
+            }
+
+            if (namespaceDeclarationCount == 0 && root is CompilationUnitSyntax compilationUnit)
+            {
+                // Should trigger if cursor is on the name of first member declaration in the 
+                // compilation unit when there's no namespace decalration in the document.
+                var firstMemberDeclaration = compilationUnit.Members.FirstOrDefault();
+                return (firstMemberDeclaration != null 
+                    && firstMemberDeclaration.GetNameToken().Span.IntersectsWith(position) 
+                    && !ContainsPartialDeclaration(document, compilationUnit, cancellationToken), null);
+            }
+
+            return (false, null);
+        }
+
+        private bool ContainsPartialDeclaration(Document document, SyntaxNode node, CancellationToken cancellationToken = default)
+        {
+            // This is just a quick check for `partial` keyword.
```

Comment:
this is dangerous, especially because it doesn't work for VB, and because we've talked about in C# about not requiring the modifier.  You should just use the semantic model to get the declarationsymbol for the types, and if any have multiple syntax locations, don't offer then.

Reply:
What do you mean by not requiring the modifier in C#?

Reply:
There have been open proposals (especially around generators) that would make it so that not every part of a partial type would need the "partial" keyword.  so best to avoid using that to make a determination.  it would simply break if we took those language changes.